Add mechanism to disable the "saved tensors hooks" feature

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85832
* #85830
* #85829
* __->__ #85553

The rationale for this is that functorch doesn't work with saved
variable hooks at the moment or checkpointing and we need some way to
disable it.

Concretely:
- there's a context manager that does the disabling
- this feature is disabled on a thread-local basis
- one can set an error message or use the default error message that
says the feature has been disabled

Since it is thread local I needed to update ATen/ThreadLocalState. To
make things nicer, this PR refactors all the "saved tensors hooks"
related TLS things into a single struct.

Test Plan:
- new test

zou3519 committed:
Add mechanism to disable the "saved tensors hooks" feature

The rationale for this is that functorch doesn't work with saved
variable hooks at the moment or checkpointing and we need some way to
disable it.

Concretely:
- there's a context manager that does the disabling
- this feature is disabled on a thread-local basis
- one can set an error message or use the default error message that
says the feature has been disabled

Since it is thread local I needed to update ATen/ThreadLocalState. To
make things nicer, this PR refactors all the "saved tensors hooks"
related TLS things into a single struct.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add mechanism to disable the "saved tensors hooks" feature"


The rationale for this is that functorch doesn't work with saved
variable hooks at the moment or checkpointing and we need some way to
disable it.

Concretely:
- there's a context manager that does the disabling
- this feature is disabled on a thread-local basis
- one can set an error message or use the default error message that
says the feature has been disabled

Since it is thread local I needed to update ATen/ThreadLocalState. To
make things nicer, this PR refactors all the "saved tensors hooks"
related TLS things into a single struct.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/SavedTensorHooks.h:
```diff
@@ -1,20 +1,49 @@
+  static void set_tls_state(const impl::SavedTensorDefaultHooksTLS& tls);
+
+  // NOTE: [Disabling SavedTensorDefaultHooks]
+  // A developer of a PyTorch feature may choose to disable SavedTensorDefault
+  // hooks, especially if their feature does not work with it. If they are
+  // disabled, then the following will raise an error:
+  // - Attempting to push_hooks
+  // - calling set_disabled(true) with a non-zero stack (from get_stack) size
```
The claim is that this is enough to disable the feature.

zou3519 committed:
Update on "Add mechanism to disable the "saved tensors hooks" feature"


The rationale for this is that functorch doesn't work with saved
variable hooks at the moment or checkpointing and we need some way to
disable it.

Concretely:
- there's a context manager that does the disabling
- this feature is disabled on a thread-local basis
- one can set an error message or use the default error message that
says the feature has been disabled

Since it is thread local I needed to update ATen/ThreadLocalState. To
make things nicer, this PR refactors all the "saved tensors hooks"
related TLS things into a single struct.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add mechanism to disable the "saved tensors hooks" feature"


The rationale for this is that functorch doesn't work with saved
variable hooks at the moment or checkpointing and we need some way to
disable it.

Concretely:
- there's a context manager that does the disabling
- this feature is disabled on a thread-local basis
- one can set an error message or use the default error message that
says the feature has been disabled

Since it is thread local I needed to update ATen/ThreadLocalState. To
make things nicer, this PR refactors all the "saved tensors hooks"
related TLS things into a single struct.

Test Plan:
- new test

[ghstack-poisoned]

albanD commented on aten/src/ATen/SavedTensorHooks.cpp:
```diff
@@ -15,6 +14,43 @@ namespace {
+  return tls.is_disabled;
+}
+
+void SavedTensorDefaultHooks::set_disabled(bool disabled) {
```
Could we pass the error message directly here?
Because otherwise the check below will never actually see the error message as it gets registered later.

zou3519 replied:
> I think we should fold the flag and message into a single field so that you can never be in a state where there is an error message but it is not disabled yet.

Sure. NB: this complicates the API, because it doesn't make sense to do `set_disabled(disabled=False, "some error message")`. So instead we need a set_disabled(error_message) and a set_enabled(). If you're OK with that then I'll go ahead with it.

soulitzer replied:
Looks like in the python context manager, the error message is set first, so it would been seen.
I'd still second combining the two APIs into one though, just to avoid any extra confusion.

soulitzer replied:
Another idea could be to just have an assert that the message is empty when disabled=False?

albanD replied:
Non-symetric API sounds ok for me.

albanD replied:
Note that set_enabled(disabled=False) raising an error if you pass the optional message is good as well.

zou3519 replied:
Okay, I'll play around with it, thanks for the feedback

albanD commented on aten/src/ATen/SavedTensorHooks.h:
```diff
@@ -1,20 +1,49 @@
   static void pop_hooks();
   static std::pair<PyObject*, PyObject*> get_hooks();
-  static void enable();
+  static void enable(); // NB: has nothing to do with disable() below.
```
Should we rename this to global_enable?

zou3519 replied:
Might need to name this something else, since we need to split set_disabled to set_disabled and set_enabled.

albanD reviewed (commented):
I think we should fold the flag and message into a single field so that you can never be in a state where there is an error message but it is not disabled yet.

zou3519 committed:
Update on "Add mechanism to disable the "saved tensors hooks" feature"


The rationale for this is that functorch doesn't work with saved
variable hooks at the moment or checkpointing and we need some way to
disable it.

Concretely:
- there's a context manager that does the disabling
- this feature is disabled on a thread-local basis
- one can set an error message or use the default error message that
says the feature has been disabled

Since it is thread local I needed to update ATen/ThreadLocalState. To
make things nicer, this PR refactors all the "saved tensors hooks"
related TLS things into a single struct.

Test Plan:
- new test

[ghstack-poisoned]

soulitzer commented on aten/src/ATen/SavedTensorHooks.cpp:
```diff
@@ -51,26 +46,26 @@ void SavedTensorDefaultHooks::set_tls_state(const impl::SavedTensorDefaultHooksT
 
-void SavedTensorDefaultHooks::enable() {
-  is_enabled = true;
+void SavedTensorDefaultHooks::lazy_initialize() {
```
question: if this function is called already lazily, does it make sense for its name to still contain "lazy", because that kind of makes me think that the function itself doesn't set the flag immediately.

zou3519 replied:
The reason why I called it `lazy_initialize()` is because `push_hooks` unconditionally calls `initialize()`. It does not make sense for something to be initialized every time it gets used, so I named it "lazy_initialize" to reflect that the initialization only happens once, on the first time "lazy initialize" is called.

That being said the lazy_initialize function does very little (it just sets a flag) and is easy to read. I can change it to initialize if you want, I don't have much of an opinion on the naming here

soulitzer commented on test/test_autograd.py:
```diff
@@ -6645,36 +6645,20 @@ def test_setting_default_saved_variable_hooks_twice_should_use_inner(self):
-
-        self.assertFalse(torch._C._autograd._saved_tensors_hooks_is_disabled())
-
+        self.assertTrue(torch._C._autograd._saved_tensors_hooks_is_enabled())
```
Let's also test the nested case?

zou3519 committed:
Update on "Add mechanism to disable the "saved tensors hooks" feature"


The rationale for this is that functorch doesn't work with saved
variable hooks at the moment or checkpointing and we need some way to
disable it.

Concretely:
- there's a context manager that does the disabling
- this feature is disabled on a thread-local basis
- one can set an error message or use the default error message that
says the feature has been disabled

Since it is thread local I needed to update ATen/ThreadLocalState. To
make things nicer, this PR refactors all the "saved tensors hooks"
related TLS things into a single struct.

Test Plan:
- new test

[ghstack-poisoned]

atalman commented:
@pytorchmergebot revert -m "Reverting since failed build-fisp-diff-linux_platform010-opt" -c ghfirst
